### DISCUSSION TITLE
Ability to export and import tags with Dashboards and Charts

### DISCUSSION THREAD

**Discussion by spens** [apache/superset#30629] [Ideas]
Is there any discussion yet on being able to export tags with superset objects?

We rely heavily on tags and would like to see that information included when exporting objects.  What is the best way to get that going?  If we have a proposal on how that functionality would work (assuming something doesn't already exist), where is the right place to get that vetted?

Thanks!

**Comment by spens** [comment]
Since posting my question, I found this [incredibly helpful video](https://preset.io/events/content-tagging-in-superset/) by @yousoph and @rusackas demoing the tagging system and describing how to go about suggesting a feature.

I don't think this is a big enough effort to warrant a SIP, but I think the first two topics in the SIP template are relevant for small changes as well so I've provided that information below. I would like to get feedback before contributing to make we aren't overlooking something important.

### Motivation
We have adopted the tagging functionality and it is working quite well for us for dashboards and and charts.  Our use case and process requires us to build dashboards on a development server, export the dashboards, commit the YAMLs to source control, and finally import the ZIPs to our production server.  We would like the tags we assign on the development server to be captured in the export so they are replicated to our production server.

### Proposed Change
UI elements are not affected by this change.  All changes are related to updated logic/processing of the existing export and import functionality.

Only tags of type "custom" are exported.  Other tags that exist but are not exported are "type" and "owner".

**Export Logic**
When an object (Dashboard/Chart) is exported, a new parameter is added to the exported YAML file.  The parameter is "tags" and the value is an array of tag names.  Here is an example:
```
slice_name: Vaccine Candidates per Approach & Stage
description: null
tags:
  - CovidTag
  - Items that start with the letter V
  - Superset Example Data
certified_by: null
...
```
Because tags have additional information (for example, a description), a new file is needed to hold this information.  The file is called "tags.yaml" and lives in the top level of the zip file.  For example:
```
dashboard_export_20241017T025146.zip
  charts (folder)
  dashboards (folder)
  databases (folder)
  datasets (folder)
  metadata.yaml (file)
  tags.yaml (file)
```

The format of the tags.yaml file is an array of tags

```
tags:
- tag_name: CovidTag
  description: Everything related to Covid
- tag_name: Items that start with the letter V
  description: Just for fun, tag Vanilla, Vikings, and anything that starts with V
- tag_name: Superset Example Data
  description: Native example data included with Superset

```

**Import Logic**
When an object is imported, the code checks to see if any tags included in the import file.
If yes, the tag is created if it does not already exist then the tag is assigned to the imported object.  If the tag already exists, it is just assigned with no changes to the tag.

Details
1. A tag can only be created if the user doing the import has permission to create tags.  Otherwise the tagging is ignored.
2. A created tag picks up the current time as the "created_on" and "changed_on" properties for the tag
3. A created tag picks up the current user as the "created_by_fk" and "changed_by_fk" 
4. A created tag picks up the description from the tags.yaml file described above
5. If a tag already exists, the description is not updated (for consistency with other export/import behavior - for example, if you import a dashboard and the chart included in the zip already exists, Superset does not update that existing chart)

Existing Superset behavior: When importing objects that already exist, the user is prompted whether it is ok to overwrite.  If no, the import process aborts.  If yes, the old object is replaced with the new object.
1. To be consistent with this behavior, when importing an object and specifying "overwrite", old tags that are no longer specified with the object are removed, new tags that didn't originally exist are created, old tags that still remain in the new set of tags are retained (descriptions are not updated for example)

_It is important to note that in this proposal, imported tags are not added to existing tags_
This is intentional because for the use case of trying to mirror the settings and properties from one system to another, you must have the ability to remove tags from the target system.  Otherwise, for example, if you rename a tag, the new name would get added and the old name would stick around which is not the desired behavior.

Thank you to anyone taking the time to read and comment!

**Comment by rusackas** [comment]
I think this is effectively closed by https://github.com/apache/superset/pull/30833 (thanks for pointing it out @Vitor-Avila). Holler if you disagree :D
